llvm/lib/CodeGen/GlobalISel/CallLowering.cpp

```diff
@@ -845,30 +898,58 @@ bool CallLowering::handleAssignments(ValueHandler &Handler,
                                      DstMPO, DstAlign, SrcMPO, SrcAlign,
                                      MemSize, VA);
         }
-        continue;
-      }
-
-      assert(!VA.needsCustom() && "custom loc should have been handled already");
-
-      if (i == 0 && !ThisReturnRegs.empty() &&
-          Handler.isIncomingArgumentHandler() &&
-          isTypeIsValidForThisReturn(ValVT)) {
+      } else if (i == 0 && !ThisReturnRegs.empty() &&
+                 Handler.isIncomingArgumentHandler() &&
+                 isTypeIsValidForThisReturn(ValVT)) {
         Handler.assignValueToReg(ArgReg, ThisReturnRegs[Part], VA);
-        continue;
-      }
-
-      if (Handler.isIncomingArgumentHandler())
+      } else if (Handler.isIncomingArgumentHandler()) {
         Handler.assignValueToReg(ArgReg, VA.getLocReg(), VA);
-      else {
+      } else {
         DelayedOutgoingRegAssignments.emplace_back([=, &Handler]() {
           Handler.assignValueToReg(ArgReg, VA.getLocReg(), VA);
         });
       }
+
+      // Finish the handling of indirect parameter passing when receiving
+      // the value (we are in the called function or the caller when receiving
+      // the return value).
+      if (VA.getLocInfo() == CCValAssign::Indirect &&
+          Handler.isIncomingArgumentHandler()) {
+        Align Alignment = DL.getABITypeAlign(Args[i].Ty);
+        MachinePointerInfo MPO;
+
+        if (VA.isMemLoc()) {
+          if (!Flags.isByRef()) {
+            LLT MemTy = Handler.getStackValueStoreType(DL, VA, Flags);
+            Handler.getStackAddress(
+              MemTy.getSizeInBytes(), VA.getLocMemOffset(), MPO, Flags);
+          } else {
+            uint64_t MemSize = Flags.getByValSize();
+            int64_t Offset = VA.getLocMemOffset();
+  
+            Handler.getStackAddress(MemSize, Offset, MPO, Flags);
```

Comment:
Do you need to use the result of `getStackAddress` at all?

Reply:
The result is not needed, but the work it does on MPO is needed.

Reply:
Should we replace this call with `MPO = MachinePointerInfo::getStack (MF, Offset`? I think that is all this function is doing to MPO, and we have `Offset` and `MF` at this function here.

Reply:
Yes I have looked at the implementation too, there are multiple targets and multiple target handlers that implement this and they all do this to set this, so it would make sense to replace it with the concrete setting of this variable, but in the current form, if a target decides to change how it wants to set the MPO it can.

The question is basically that, whether in the future will there be a target, that needs to set MPO differently for the stack address?

I don't know how likely it is that there will be.

Reply:
One solution to avoid the creation of dead instructions is to add a `getMachinePointerInfoForStackAddress` function that targets can override we call here and in `getStackAddress`.

Reply:
but we could probably leave that as a TODO

Reply:
Now I have looked more into this whole stack MPO creation I don't think it could change that much. It would be the best, that it should not be constructed individually by every implementer of the handler, but rather there would be a wrapper before the actual call that construct this.

There are already some cases when there is a non virtual wrapper before a handler call.

Reply:
For example `assignValueToAddress` has this non virtual wrapper wrapper:
```cpp
virtual void assignValueToAddress(const ArgInfo &Arg, unsigned ValRegIndex,
                                      Register Addr, LLT MemTy,
                                      const MachinePointerInfo &MPO,
                                      const CCValAssign &VA) {
      assignValueToAddress(Arg.Regs[ValRegIndex], Addr, MemTy, MPO, VA);
    }
```
To get the concrete register out with and index.